### PR TITLE
Implement reveal

### LINKED ISSUE #393: An error occurs when clicking "Click to view resource" without Azure view opening

### PR DESCRIPTION
**Pull request by alexweininger**
We have to do some additional acrobatics 🥷  to make reveal work consistently.

For one, if `onDidChangeTreeData` is fired while reveal is executing, reveal will fail. When `onDidChangeTreeData` is fired, VS Code's internal cache is cleared right before it attempts to reveal, and reveal fails since VS Code no longer has a reference to the element to reveal. This might be a VS Code issue, but it's fixable for now by blocking firing that event until reveal has finished executing.

I've implemented so that `onDidChangeTreeData` events fired while reveal is executing are deferred until reveal finishes.

We expose application and workspace resource tree views to the client extensions. Client extensions use `TreeView.reveal` directly. To apply this custom logic around reveal, I had to directly modify the `TreeView.reveal` method.

Fixes #393 🥳 

Also, having a little dilemma...

Find tree item is super slow if it calls `getChildren` instead of getting the children from the cache. This results in reveal being super slow too since they work in tandem in many cases. I measured reveal taking 2s when using the cache, and 15s without using it. 15s is long enough users will assume the operation failed IMO.

Well Alex why don't you just use get the children from the cache then?... 😄

A few issues arise:

A. Fetching the children from the cache causes #416. When the cache isn't used, #416 does not reproduce.

B. Ideally the only thing find tree item needs to work is a `vscode.TreeDataProvider<ResourceGroupsItem>`. If we need to use the cache in find tree item, then we'll need a reference to it and expose that in the API. We'd like to avoid this as we want to keep the API as simple as possible. And as close to VS Code's API as possible.

For now I'm using the cache because I don't think we can allow reveal to be that slow, but I'm open to opinions and suggestions here.

### CODE REVIEW

File: src/tree/v2/application/registerResourceGroupsTreeV2.ts [hunk 1]
```diff
@@ -35,21 +36,23 @@ export function registerApplicationTree(context: vscode.ExtensionContext, option
 
     context.subscriptions.push(resourceGroupingManager);
 
-    const applicationResourceTreeDataProvider = new ApplicationResourceTreeDataProvider(branchDataProviderManager.onDidChangeTreeData, itemCache, refreshEvent, resourceGroupingManager, resourceProviderManager);
+    const applicationResourceTreeDataProvider = new ApplicationResourceTreeDataProvider(
+        branchDataProviderManager.onDidChangeTreeData,
+        itemCache, refreshEvent,
```
Comment (bwateratmsft) [inline src/tree/v2/application/registerResourceGroupsTreeV2.ts:41]:
```suggestion
        itemCache,
        refreshEvent,
```

**Comment by bwateratmsft** [comment]
> Find tree item is super slow if it calls getChildren instead of getting the children from the cache. This results in reveal being super slow too since they work in tandem in many cases. I measured reveal taking 2s when using the cache, and 15s without using it. 15s is long enough users will assume the operation failed IMO.

🤮 

@alexweininger we need to find out what the heck is taking so long. I'd expect two seconds to be enough time to enumerate the whole tree top to bottom. Fifteen is just unacceptable.

**Comment by alexweininger** [comment]
> 🤮
> 
> @alexweininger we need to find out what the heck is taking so long. I'd expect two seconds to be enough time to enumerate the whole tree top to bottom. Fifteen is just unacceptable.

I have 15 App Service apps in my subscription, each App Service takes ~700ms to resolve (GET request). That's at least 10s. Also I think 15 resources is a realistic number.

**Comment by bwateratmsft** [comment]
Agreed, 15 resources is reasonable. We need to make some changes then. The calls should be able to be "parallel"ized (inasmuch as that's a thing when there's only one thread). That should help a lot. Additionally, the less we can do during resolve, the better.

**Comment by alexweininger** [comment]
> Agreed, 15 resources is reasonable. We need to make some changes then. The calls should be able to be "parallel"ized (inasmuch as that's a thing when there's only one thread). That should help a lot since these things are all network-bound, not CPU-bound. Additionally, the less we can do during resolve, the better.

For a group, we're already using `Promise.all` for concurrency when calling `getResourceItem`. So I don't think there's much more optimization to do there.

However, inside of `getResourceItem` instead of making individual GET requests we can make GET request to list resources, which should help significantly for when groups have multiple resources of the same type.

We could change `getResourceItem` from 

```ts
getResourceItem(element: TResource): TModel | Thenable<TModel>;
```

to 

```ts
getResourceItem(elements: TResource[]): TModel[] | Thenable<TModel[]>;
```

We could also make list the resources on the first `getResourceItem` call, and then use the cached result on subsequent calls. However other than setting an arbitrary cache invalidation time, I'm not sure how to properly invalidate this cache.

**Comment by bwateratmsft** [comment]
I don't want to change `getResourceItem`. I'd rather fix the client extensions to be less awful. Though I'm surprised that even with concurrency it's taking this long.

**Comment by alexweininger** [comment]
Converting to draft until find is finalized.

**Comment by alexweininger** [comment]
Closing as stale and replaced by newer PRs

**Pull request closed without merging**